src/core-services/orders/queries/ordersByAccountId.js

```diff
@@ -42,10 +42,8 @@ export default async function ordersByAccountId(context, { accountId, orderStatu
       const shopIdsUserHasPermissionFor = shopsUserHasPermissionFor("orders");
       query.shopId = { $in: shopIdsUserHasPermissionFor };
     } else {
-      shopIds.forEach((shopId) => {
-        if (!userHasPermission(["orders", "order/fulfillment"], shopId)) {
-          throw new ReactionError("access-denied", "Access Denied");
-        }
+      shopIds.forEach(async (shopId) => {
+        await checkPermissions(["orders", "order/fulfillment"], shopId);
```

Comment:
Is this OK to do in a `forEach`, or does it need to be done in a `for...of` instead?

Reply:
You cannot just slap an `async` on a forEach function and expect it to work how you want, sorry :(. You should use a `for-of` loop with an `await` inside the block. We also have [`for-await-of`](https://node.green/#ES2018-features-Asynchronous-Iterators-for-await-of-loops) as of NodeJS 10.x.

In general I like to use normal for/for-in/for-of loops everywhere unless there is some situation in which a forEach actually gives us something useful (which is typically just so we don't have to keep an incrementing index variable, but that's pretty rare and thus so are forEach's). Even in such cases you have to be careful, as errors can get swallowed and stack traces can be confusing.

Reply:
👍 All these instances have already been refactored with a `for...of`. I thought that was the case when I was refactoring the code, figured with a `forEach` it would skip past it while it was running, but just wanted to double check before making the updates.